Fix rtmp secret validation to allow `/` (#1069)

jeyemwey opened the pull request:
Now allowing `/` in the stream secret.

This does rely on the middle part being `/live/` which was not enforced before. All of our docs use `/live/` in the examples, but there was nothing that stopped users to set this to other values.

Tangentely related to #1069 which reported that they're unable to use `/` secrets in OBS.

volkland committed:
Fix rtmp secret validation to allow `/` (#1069)

aaronpk reviewed (changes requested):
Can you add some more failure tests, things like

* `rtmp://example.com/one/two/live/three`
* `rtmp://example.com/streamkey/live`
* `rtmp://example.com/somethingelse/streamkey`
*

volkland committed:
add negative test cases for stuff before /live/

jeyemwey commented:
:heavy_check_mark:

Added some more tests and added a test if there's stuff between the hostname and `/live/`.

volkland committed:
simplify since Url.Path is already stripping the host

This means that we can simplify the code and make it much clearer.
Removes the tests that checked for the host and stuff between the host and /live/.

jeyemwey commented:
The integration tests on 85601a1 showed a massive flaw in my function, namely the schema and the domain were only transmitted to the function in my tests, because `c.URL.Path` [is not what I thought it was](https://pkg.go.dev/net/url#URL.Path).

Anyways, the code is much clearer now and all the tests are happy, too.

gabek commented:
Could somebody catch me up?  I'm mis-interpreting this issue.  The original issue said that Owncast was fine, OBS doesn't support slashes so we should put something in the admin to let people know not to use them.   The reason I closed it was it was said there was not a technical problem, but there is?  Sorry if I'm not following what's going on here.

aaronpk commented:
The problem is that OwnCast is currently incorrectly parsing the RTMP URL when validating the stream key. This problem is not anything to do with OBS.

If the stream key in owncast is set to `one/two/three`, the client would send a URL like `rtmp://example.com/live/one/two/three` and the current code at line 81 (splitting on the slashes) would be grabbing just `three` and comparing it to `one/two/three` and reject the stream.

Another separate problem is OwnCast is essentially ignoring the first part of the RTMP path, `live`, so it will accept an RTMP URL with anything in that segment.

This PR fixes both issues by parsing the RTMP URL better when checking for the stream key.

gabek commented:
Thanks for fixing this!  Sorry for my mis-reading of the original issue.
